File: 03.py
import sys


def multiply_order(p):
    n = len(p)
    dp = [[0] * n for i in range(n)]
    for l in range(1, n - 1):
        for i in range(1, n - l):
            j = i + l
            dp[i][j] = sys.maxsize
            for k in range(i, j):
                dp[i][j] = min(dp[i][j], dp[i][k] + dp[k + 1][j] + p[i - 1] * p[k] * p[j])
    return dp[1][n - 1]


with open("input.txt", "r") as fin:
    _n_ = int(fin.readline())
    dims = []
    for line in fin.readlines():
        _n, _m = map(int, line.strip().split())
        if _n not in dims:
            dims.append(_n)
        if _m not in dims:
            dims.append(_m)

with open("output.txt", "w") as fout:
    operations = multiply_order(dims)
    fout.write(str(operations))

File: 03.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <algorithm>
#include <limits>

int multiply_order(std::vector<int> p) {
    int n = p.size();
    int** dp = new int*[n];
    for (int i = 0; i < n; i++) {
        dp[i] = new int[n];
    }

    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            dp[i][j] = 0;
        }
    }

    for (int l = 1; l < n - 1; l++) {
        for (int i = 1; i < n - l; i++) {
            int j = i + l;
            dp[i][j] = std::numeric_limits<int>::max();

            for (int k = i; k < j; k++) {
                dp[i][j] = std::min(dp[i][j],
                                    dp[i][k] + dp[k + 1][j] + p[i - 1] * p[k] * p[j]);
            }
        }
    }

    int k = dp[1][n - 1];

    for (int i = 0; i < n; i++) {
        delete dp[i];
    }
    delete[] dp;

    return k;
}

int main(int argc, char const *argv[]) {
    std::ifstream fin("input.txt");
    int d;
    fin >> d;

    std::vector<int> dims;
    fin >> d;
    dims.push_back(d);
    while (!fin.eof()) {
        fin >> d;
        dims.push_back(d);
        fin >> d;
    }
    fin.close();

    int k = multiply_order(dims);
    std::ofstream fout("output.txt");
    fout << k << std::endl;
    fout.close();

    return 0;
}
